#Meta-omics scripts for processing and analysis of metagenomic and metatranscriptomic data

Scripts for processing read, mapping, annotation, and coverage data from large meta-omic runs.

Programming Libraries used:   
Python 2.7.9  - Numpy 1.9.1, Matplotlib 1.3.1, Biopython 1.64, Pandas 0.15.1   
Perl 5.16.3   
R 3.1.1 - RStudio 0.98.1049, vegan 2.2-1, RColorBrewer 1.1-2, alphahull 1.0, e1071 1.6-4, fpc 2.1-9, rgl 0.93.1098, fields 7.1    
bamtools 2.20   
bcftools 1.2   
bedtools v2.24.0   
bowtie2 2.2.4   
freebayes v0.9.21   
Hmmsearch, HMMR 3.1b1   
Prodigal 2.6.1   
samtools 0.1.19    
Bash shell scripting (NERSC compatible scripts included) - scp, tar, cat, cut, head, tail, paste, sed, unique, grep, vim NOTE: syntax will vary depending on platform and version.  For Macs, gsed could be more useful than the Darwin sed and is used in this guide.


Version 1.0

The MIT License (MIT)
https://opensource.org/licenses/MIT
Copyright (c) 2015 Jackson Lee

gitub: leejz
